libvast/vast/system/spawn_source.hpp

```diff
@@ -14,50 +14,59 @@
-/// @param args Configures the new actor.
-/// @returns a handle to the spawned actor on success, an error otherwise
-maybe_actor spawn_suricata_source(node_actor* self, spawn_arguments& args);
-
-/// Tries to spawn a new SOURCE for the Syslog format.
-/// @param self Points to the parent actor.
-/// @param args Configures the new actor.
-/// @returns a handle to the spawned actor on success, an error otherwise
-maybe_actor spawn_syslog_source(node_actor* self, spawn_arguments& args);
-
-/// Tries to spawn a new SOURCE for generating random data.
-/// @param self Points to the parent actor.
-/// @param args Configures the new actor.
-/// @returns a handle to the spawned actor on success, an error otherwise
-maybe_actor spawn_test_source(node_actor* self, spawn_arguments& args);
-
-/// Tries to spawn a new SOURCE for the Zeek format.
-/// @param self Points to the parent actor.
-/// @param args Configures the new actor.
-/// @returns a handle to the spawned actor on success, an error otherwise
-maybe_actor spawn_zeek_source(node_actor* self, spawn_arguments& args);
+template <class Reader, class Defaults = typename Reader::defaults>
+maybe_actor spawn_source(node_actor* self, spawn_arguments& args) {
+  VAST_TRACE(VAST_ARG("node", self), VAST_ARG(args));
+  auto& st = self->state;
+  auto& options = args.inv.options;
+  // Bail out early for bogus invocations.
+  if (caf::get_or(options, "system.node", false))
+    return make_error(ec::parse_error, "cannot start a local node");
+  VAST_UNBOX_VAR(in, detail::make_input_stream<Defaults>(args.inv.options));
+  VAST_UNBOX_VAR(sch, read_schema(args));
+  auto table_slice_type = caf::get_or(options, "source.spawn.table-slice-type",
+                                      defaults::import::table_slice_type);
+  auto slice_size = get_or(options, "source.spawn.table-slice-size",
+                           defaults::import::table_slice_size);
+  auto max_events = caf::get_if<size_t>(&options, "source.spawn.max-events");
+  auto type = caf::get_if<std::string>(&options, "source.spawn.type");
+  auto type_filter = type ? std::move(*type) : std::string{};
+  auto schema = get_schema(options, "spawn.source");
+  if (!schema)
+    return schema.error();
+  if (slice_size == 0)
+    return make_error(ec::invalid_configuration, "table-slice-size can't be 0");
+  Reader reader{table_slice_type, options, std::move(in)};
+  VAST_INFO(self, "spawned a", reader.name(), "source");
+  auto src
+    = self->spawn<caf::detached>(source<Reader>, std::move(reader), slice_size,
+                                 max_events, st.type_registry, vast::schema{},
+                                 std::move(type_filter), accountant_type{});
+  src->attach_functor([=, name = reader.name()](const caf::error& reason) {
+    if (!reason || reason == caf::exit_reason::user_shutdown)
+      VAST_INFO(name, "source shuts down");
+    else
+      VAST_WARNING(name, "source shuts down with error:", reason);
```

Comment:
```suggestion
      VAST_WARNING(name, "source shut down with error:", reason);
```

Reply:
But this is before it has completely shut down, so I think present tense is okay here?

Reply:
My understanding was that `attach_functor` executes after the actor terminated. But [looking at the docs](https://actor-framework.readthedocs.io/en/latest/Actors.html#attaching-cleanup-code-to-actors), all it says is *during*. Since you can attach cleanup code for remote actors as well, it's equivalent to receiving a DOWN message. Therefore, I suggest we treat `attach_functor` as executed after shutdown.

Reply:
Addressed.